src/molinillo_solver.cr

```diff
@@ -21,27 +21,15 @@ module Shards
 
     private def add_lock(base, lock_index, dep : Dependency)
       if lock = lock_index.delete(dep.name)
-        lock_version =
-          case lock_req = lock.requirement
-          when Version then lock_req
-          else
-            versions = lock.resolver.versions_for(lock_req)
-            unless versions.size == 1
-              Log.warn { "Lock for shard \"#{dep.name}\" is invalid" }
-              return
-            end
-            lock.requirement = versions.first
-          end
-
         check_single_resolver_by_name dep.resolver
-        base.add_vertex(lock.name, lock, true)
+        base.add_vertex(lock.name, Dependency.new(lock.name, lock.resolver, lock.version), true)
```

Comment:
```suggestion
        base.add_vertex(lock.name, Dependency.new(lock.name, dep.resolver, lock.version), true)
```

We should use dep.resolver to allo chaning source without bumping the version when possible.
Otherwise `check_single_resolver_by_name` should be called with lock.resolver and it will get ambiguos source when the shard.yml and lock differ in source.

Reply:
Is that a scenario not already covered by specs?

Reply:
It was added in https://github.com/crystal-lang/shards/pull/419/files#diff-d38f7bcbd4d1fda0933136e184879150R268-R282 . With your proposed change (using `lock.resolver`) I would expect that the `assert_locked` would wrongly use `git_url(:awesome)` instead of `git_url(:forked_awesome)`

Reply:
Ok, since previously the whole lock was added as a vertex and that included the lock's resolver.

Reply:
Anyway I changed the resolver as you suggested. It's the right thing, even though the resolver from the locks is not being used in the resolution process.